workspace_server/src/main/java/gov/nasa/jpl/aerie/workspace/server/WorkspaceFileSystemService.java

```diff
@@ -159,6 +160,98 @@ public boolean moveFile(final int workspaceId, final Path oldFilePath, final Pat
+  public boolean moveFile(final int workspaceId, final Path oldFilePath, final Path newFilePath)
+  throws NoSuchWorkspaceException, SQLException, WorkspaceFileOpException
+  {
+    return moveFile(workspaceId, oldFilePath, workspaceId, newFilePath);
+  }
+
+  @Override
+  public boolean copyFile(final int sourceWorkspaceId, final Path sourceFilePath, final int destWorkspaceId, final Path destFilePath)
+  throws NoSuchWorkspaceException, SQLException, WorkspaceFileOpException
+  {
+    final var sourceRepoPath = postgresRepository.workspaceRootPath(sourceWorkspaceId);
+    final var sourcePath = sourceRepoPath.resolve(sourceFilePath);
+    final var destRepoPath = postgresRepository.workspaceRootPath(destWorkspaceId);
+    final var destPath = destRepoPath.resolve(destFilePath);
+
+    try {
+      // Do not copy the file if the source file does not exist
+      if(!sourcePath.toFile().exists()) throw new WorkspaceFileOpException("Source file \"%s\" in workspace %d does not exist.".formatted(sourceFilePath, sourceWorkspaceId));
+
+      // Do not copy the file if the destination already exists
+      if(destPath.toFile().exists()) throw new WorkspaceFileOpException("Destination file \"%s\" in workspace %d already exists.".formatted(destFilePath, destWorkspaceId));
+
+      // Create any parent directories that don't already exist
+      Files.createDirectories(destPath.getParent());
+
+      // Copy the main file
+      Files.copy(sourcePath, destPath);
+
+      // Find and copy hidden metadata files
+      final var metadataExtensions = postgresRepository.getMetadataExtensions();
+      for (final var extension : metadataExtensions) {
+        final var oldMetaPath = Path.of(sourcePath + extension);
+        final var newMetaPath = Path.of(destPath + extension);
+        if (Files.exists(oldMetaPath)) {
+          Files.copy(oldMetaPath, newMetaPath);
+        }
+      }
+
+      return true;
+    } catch (IOException e) {
+      e.printStackTrace();
+      return false;
+    }
+  }
+
+
+  @Override
+  public boolean copyDirectory(final int sourceWorkspaceId, final Path sourceFilePath, final int destWorkspaceId, final Path destFilePath)
+  throws NoSuchWorkspaceException, WorkspaceFileOpException
+  {
+    final var sourceRepoPath = postgresRepository.workspaceRootPath(sourceWorkspaceId);
+    final var sourcePath = sourceRepoPath.resolve(sourceFilePath);
+    final var destRepoPath = postgresRepository.workspaceRootPath(destWorkspaceId);
+    final var destPath = destRepoPath.resolve(destFilePath);
+
+    try {
+      // Validate source exists and is a directory
+      if (!Files.exists(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d does not exist.".formatted(sourceFilePath, sourceWorkspaceId));
+      if (!Files.isDirectory(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d is not actually a directory. This is a bug.".formatted(sourceFilePath, sourceWorkspaceId));
```

Comment:
Please exclude the phrase "This is a bug" from the error message.
```suggestion
      if (!Files.isDirectory(sourcePath)) throw new WorkspaceFileOpException("Source directory \"%s\" in workspace %d is not actually a directory.".formatted(sourceFilePath, sourceWorkspaceId));
```

Additionally, this state is not unsalvageable -> you could just go into the `copyFile` with the same args

Reply:
Fixed in ab8d781, I think it's OK to throw in this "shouldn't happen" case so we can figure out why it did